zram-swap: Support more zram config options

vortigont opened the pull request:
+ "compression algorithm" configuration
  Compression algorithms for zram are provided by kernel crypto API, could be
  any of [lzo|zl4|deflate|<some_more>] depending on kernel modules.
  Compress algo for zram-swap could be defined via 'zram_comp_algo' option in
  'system' section of '/etc/config/system' file,
  or via cli (for e.x. with 'uci set system.@system[0].zram_comp_algo=lz4 && uci commit system').
  check available algo's via 'cat /sys/block/zram0/comp_algorithm'

+ "max compression streams" configuration
  Config option to limit maximum compression streams for zram dev for multicore CPU's.
  This could be defined via 'zram_comp_streams' option in the 'system' section of '/etc/config/system' file
  or via cli (for e.x. with 'uci set system.@system[0].zram_comp_streams=2 && uci commit system').
  Default is number of logical CPU cores.

* fix: use only one zram swap device of the specified size instead of [N x $size] devices for multicore CPUs
  Now zram module uses multiple compression streams for each dev, so we do not need
  to create several zram devs to utilize multicore CPUs.

* fix: zram dev reset for multicore cpu devices
  "zram stop" could reset up to $(num_of_cores) zram devices even if some of those
  were not mounted as swap dev's. This fix tries to enumerate only swap zram dev's before making reset

* fix: hot-add new zram dev for swap if zram0 is already mounted (for ex. /tmp)

* remove hot-added zram devs on stop

tested on:
 - trunk mt7621 kern 4.14.32 multicore
 - lede 17.1 ar71 kern 4.4.92 single core
 - CC 15.05 ar71 kern 3.18.20 single core

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

blogic commented:
@vortigont looks sort of ok but its really hard to review with all the changes in 1 single patch. please split this up into 1 patch per change that you make.

vortigont committed:
zram-swap: compression algorithm configuration option
      Compression algorithms for zram are provided by kernel crypto API, could be
      any of [lzo|zl4|deflate|<some_more>] depending on kernel modules.
      Compress algo for zram-swap could be defined via 'zram_comp_algo' option in
      'system' section of '/etc/config/system' file,
      or via cli (for e.x. with 'uci set system.@System[0].zram_comp_algo=lz4 && uci commit system').
      check available algo's via 'cat /sys/block/zram0/comp_algorithm'

  Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
zram-swap: fix zram dev reset for multicore cpu devices
  * "zram stop" could reset up to $(num_of_cores) zram devices even if some of those
      were not mounted as swap dev's.
	This fix tries to enumerate mounted swap zram dev's before making a reset

  * remove hot-added zram devs on stop (except zram0)

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
zram-swap: fix number of created zram devices for multicore CPU's
  use only one zram swap device of the specified $size instead of [N x $size] devices for multicore CPUs
  Now zram module uses multiple compression streams for each dev by default,
  so we do not need to create several zram devs to utilize multicore CPUs.

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
zram-swap: Add "max compression streams" configuration option
  Config option to limit maximum compression streams per zram dev for multicore CPU's.
  This could be defined via 'zram_comp_streams' option in the 'system' section of '/etc/config/system' file
  or via cli (for e.x. with 'uci set system.@System[0].zram_comp_streams=2 && uci commit system').
  Default is number of logical CPU cores.

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont commented:
@jsmcconn, any chance to look into this? tnx

jsmcconn commented:
After reading through the threads I observe the following:

* This issue is exposed only in intensive swap situations on 4.15+ kernels, when ZRAM_WRITEBACK is enabled.
* The issue has been patched in upstream 4.19, and 4.18.7, but 4.14.y has not been updated.

Currently openwrt doesn't support kernels newer than 4.14.y nor does any configuration enable ZRAM_WRITEBACK, which is disabled by default.

**My suggestion is to take no action and let upstream handle the issue.**

Anecdotally, my mvebu system actively using this feature has a 70+ day uptime with no oopses,  although the system seldom if ever gets into swap.

vortigont commented:
tnx @jsmcconn. Agreed, zram fixes if any has nothing to do with a script here. Usually openwrt platforms doesn't have any backing block device to use for swapping (at least talking about routers), nor this script offers any way to configure it. My intention was to update rather old init functions to fit some features, that has been added to zram kernel module since 3.x.
Although maybe @hmh meant to consider implementing backing store configuration option too? :)

hmh commented:
Nah, I was just pointing out those two issues needed to be looked into to know whether they would apply or not to OpenWRT.  If they don't, case closed, full speed ahead :-)

vortigont committed:
zram-swap: bump pkg version

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

hauke commented:
Thank you for the patch it was applied to master.
